File: src/pulsemesh/local.py
from __future__ import annotations

import os
import platform
import re
import shutil
import socket
import subprocess
import time

from .models import TelemetryProfile, TelemetrySeries


def fetch_system(profile: TelemetryProfile, max_points: int, _: float) -> TelemetrySeries:
    samples = max(4, min(max_points, int(profile.params.get("samples", 24))))
    interval = max(0.0, float(profile.params.get("interval", 0.0)))
    variable = (profile.variable or str(profile.params.get("variable", "cpu_load"))).lower()
    values: list[float] = []
    times: list[str] = []

    for i in range(samples):
        if variable in {"disk_free_percent", "disk_free"}:
            total, used, free = shutil.disk_usage(str(profile.params.get("path", ".")))
            value = 100.0 * free / total
        elif variable in {"disk_used_percent", "disk_used"}:
            total, used, free = shutil.disk_usage(str(profile.params.get("path", ".")))
            value = 100.0 * used / total
        elif variable in {"load1", "cpu_load"}:
            try:
                value = float(os.getloadavg()[0])
            except (AttributeError, OSError):
                value = _windows_cpu_proxy()
        elif variable in {"process_count", "processes"}:
            value = float(_process_count())
        elif variable in {"memory_used_percent", "memory"}:
            value = _windows_numeric("(Get-CimInstance Win32_OperatingSystem | ForEach-Object { (($_.TotalVisibleMemorySize - $_.FreePhysicalMemory) / $_.TotalVisibleMemorySize) * 100 })")
        elif variable in {"memory_free_percent"}:
            value = _windows_numeric("(Get-CimInstance Win32_OperatingSystem | ForEach-Object { ($_.FreePhysicalMemory / $_.TotalVisibleMemorySize) * 100 })")
        elif variable in {"battery_percent", "battery"}:
            value = _windows_numeric("(Get-CimInstance Win32_Battery | Measure-Object -Property EstimatedChargeRemaining -Average).Average")
        elif variable in {"uptime_hours", "uptime"}:
            value = _uptime_hours()
        elif variable in {"net_bytes_sent", "network_sent"}:
            value = float(_network_bytes(sent=True))
        elif variable in {"net_bytes_recv", "network_recv"}:
            value = float(_network_bytes(sent=False))
        else:
            raise ValueError(f"unknown system variable: {variable}")
        values.append(value)
        times.append(f"sample:{i:04d}")
        if interval > 0 and i != samples - 1:
            time.sleep(interval)

    return TelemetrySeries(
        profile_id=profile.id,
        provider=profile.provider,
        label=profile.label or f"System {variable}",
        sensor_name=f"Local system sensor: {platform.node()}",
        times=times,
        values=values,
        unit="percent" if "percent" in variable or variable.startswith("disk_") else "count",
        metadata={"variable": variable, "platform": platform.platform()},
    )


def fetch_ping(profile: TelemetryProfile, max_points: int, timeout: float) -> TelemetrySeries:
    host = str(profile.params.get("host") or profile.params.get("target") or "1.1.1.1")
    count = max(4, min(max_points, int(profile.params.get("count", 8))))
    values: list[float] = []
    times: list[str] = []
    for i in range(count):
        started = time.perf_counter()
        ok = _tcp_probe(host, int(profile.params.get("port", 443)), timeout)
        elapsed_ms = (time.perf_counter() - started) * 1000.0
        values.append(elapsed_ms if ok else timeout * 1000.0)
        times.append(f"probe:{i:04d}")
    return TelemetrySeries(
        profile_id=profile.id,
        provider=profile.provider,
        label=profile.label or f"TCP latency {host}",
        sensor_name=f"TCP probe {host}:{profile.params.get('port', 443)}",
        times=times,
        values=values,
        unit="ms",
        metadata={"host": host, "port": int(profile.params.get("port", 443))},
    )


def _tcp_probe(host: str, port: int, timeout: float) -> bool:
    try:
        with socket.create_connection((host, port), timeout=timeout):
            return True
    except OSError:
        return False


def _windows_cpu_proxy() -> float:
    try:
        cmd = ["powershell", "-NoProfile", "-Command", "(Get-CimInstance Win32_Processor | Measure-Object -Property LoadPercentage -Average).Average"]
        out = subprocess.check_output(cmd, text=True, timeout=5).strip()
        return float(out or 0.0)
    except Exception:
        return 0.0


def _process_count() -> int:
    try:
        cmd = ["powershell", "-NoProfile", "-Command", "(Get-Process).Count"]
        out = subprocess.check_output(cmd, text=True, timeout=5).strip()
        return int(float(out or 0))
    except Exception:
        return 0


def _windows_numeric(script: str) -> float:
    try:
        cmd = ["powershell", "-NoProfile", "-Command", script]
        out = subprocess.check_output(cmd, text=True, timeout=5).strip()
        match = re.search(r"-?\d+(?:\.\d+)?", out)
        return float(match.group(0)) if match else 0.0
    except Exception:
        return 0.0


def _uptime_hours() -> float:
    try:
        return float(time.monotonic() / 3600.0)
    except Exception:
        return 0.0


def _network_bytes(sent: bool) -> int:
    try:
        prop = "BytesSentPersec" if sent else "BytesReceivedPersec"
        script = f"(Get-CimInstance Win32_PerfFormattedData_Tcpip_NetworkInterface | Measure-Object -Property {prop} -Sum).Sum"
        return int(_windows_numeric(script))
    except Exception:
        return 0

File: agnt-plugin/local.js
// PulseMesh JS — Local System Telemetry
import os from 'os';
import fs from 'fs';
import { spawnSync } from 'child_process';
import { TelemetrySeries } from './models.js';
import { nowISO } from './util.js';

function currentTimes(n) {
  const base = Date.now();
  const step = 1000;
  return Array.from({ length: n }, (_, i) => new Date(base - (n - 1 - i) * step).toISOString().replace(/\.\d+Z$/, 'Z'));
}

function getCpuLoad() {
  try {
    if (process.platform === 'win32') {
      const r = spawnSync('powershell', ['-Command',
        '(Get-CimInstance Win32_Processor | Measure-Object -Property LoadPercentage -Average).Average'
      ], { timeout: 3000 });
      if (r.status === 0 && r.stdout) {
        const v = parseFloat(r.stdout.toString().trim());
        if (Number.isFinite(v)) return v;
      }
      const cpus = os.cpus();
      let totalIdle = 0, totalTick = 0;
      for (const c of cpus) { totalIdle += c.times.idle; totalTick += Object.values(c.times).reduce((a, b) => a + b, 0); }
      return totalTick > 0 ? 100 * (1 - totalIdle / totalTick) : 0;
    }
    return (os.loadavg()[0] / os.cpus().length) * 100;
  } catch { return 0; }
}

function getMemoryUsedPercent() {
  try {
    if (process.platform === 'win32') {
      const r = spawnSync('powershell', ['-Command',
        '((Get-CimInstance Win32_OperatingSystem).TotalVisibleMemorySize - (Get-CimInstance Win32_OperatingSystem).FreePhysicalMemory) / (Get-CimInstance Win32_OperatingSystem).TotalVisibleMemorySize * 100'
      ], { timeout: 3000 });
      if (r.status === 0) { const v = parseFloat(r.stdout.toString().trim()); if (Number.isFinite(v)) return v; }
    }
    const total = os.totalmem(); const free = os.freemem();
    return 100 * (1 - free / total);
  } catch { return 0; }
}

function getMemoryFreePercent() { return 100 - getMemoryUsedPercent(); }

function getDiskFreePercent(drive = '.') {
  try {
    if (process.platform === 'win32') {
      const letter = (drive === '.' ? process.cwd()[0] : drive[0]).toUpperCase();
      const r = spawnSync('powershell', ['-Command',
        `(Get-PSDrive ${letter} -ErrorAction SilentlyContinue).Free / ((Get-PSDrive ${letter} -ErrorAction SilentlyContinue).Free + (Get-PSDrive ${letter} -ErrorAction SilentlyContinue).Used) * 100`
      ], { timeout: 3000 });
      if (r.status === 0) {
        const v = parseFloat(r.stdout.toString().trim());
        if (Number.isFinite(v) && v >= 0 && v <= 100) return v;
      }
    }
    const r = spawnSync('df', ['-P', drive], { timeout: 3000 });
    if (r.status === 0) {
      const lines = r.stdout.toString().trim().split('\n');
      if (lines.length >= 2) {
        const parts = lines[1].split(/\s+/);
        const total = parseFloat(parts[1]); const used = parseFloat(parts[2]);
        if (total > 0) return 100 * (1 - used / total);
      }
    }
    return 50;
  } catch { return 0; }
}

function getProcessCount() {
  try {
    if (process.platform === 'win32') {
      const r = spawnSync('powershell', ['-Command', '(Get-Process | Measure-Object).Count'], { timeout: 3000 });
      if (r.status === 0) { const v = parseInt(r.stdout.toString().trim()); if (Number.isFinite(v)) return v; }
    }
    const r = spawnSync('ps', ['-e', '--no-headers'], { timeout: 3000 });
    if (r.status === 0) return r.stdout.toString().trim().split('\n').filter(l => l.trim()).length;
    return 0;
  } catch { return 0; }
}

function getUptime() {
  try {
    if (process.platform === 'win32') {
      const r = spawnSync('powershell', ['-Command',
        '(Get-Date) - (Get-CimInstance Win32_OperatingSystem).LastBootUpTime | Select-Object -ExpandProperty TotalHours'
      ], { timeout: 3000 });
      if (r.status === 0) { const v = parseFloat(r.stdout.toString().trim()); if (Number.isFinite(v)) return v; }
    }
    return os.uptime() / 3600;
  } catch { return 0; }
}

// Ping: uses native OS ping command for reliable RTT measurements
function getPingMs(host, count) {
  const results = [];
  try {
    if (process.platform === 'win32') {
      const r = spawnSync('ping', ['-n', String(count), '-w', '2000', host], { timeout: 15000, encoding: 'utf-8' });
      // Parse round-trip times: "time=<ms>ms" or "time<ms>ms"
      const output = r.stdout || '';
      const matches = output.matchAll(/time[=<](\d+(?:\.\d+)?)\s*ms/gi);
      for (const m of matches) { const v = parseFloat(m[1]); if (Number.isFinite(v) && v > 0) results.push(v); }
    } else {
      const r = spawnSync('ping', ['-c', String(count), '-W', '2', host], { timeout: 15000, encoding: 'utf-8' });
      const output = r.stdout || '';
      const matches = output.matchAll(/time=(\d+(?:\.\d+)?)\s*ms/gi);
      for (const m of matches) { const v = parseFloat(m[1]); if (Number.isFinite(v) && v > 0) results.push(v); }
    }
  } catch { /* return empty */ }
  return results;
}

export function fetchSystem(profile, maxPoints = 512, _timeout = 12) {
  const variable = profile.variable || profile.params?.variable || 'cpu_load';
  const samples = profile.samples || Math.min(maxPoints, 24);
  const variableMap = {
    'cpu_load': getCpuLoad,
    'memory_used_percent': getMemoryUsedPercent,
    'memory_free_percent': getMemoryFreePercent,
    'disk_free_percent': () => getDiskFreePercent(profile.params?.path || profile.path || '.'),
    'process_count': getProcessCount,
    'uptime': getUptime,
  };
  const fn = variableMap[variable] || getCpuLoad;
  const values = [];
  for (let i = 0; i < samples; i++) {
    const raw = fn();
    const val = Number.isFinite(raw) ? raw : null;
    if (val !== null) values.push(val);
  }
  const unitMap = {
    'cpu_load': '%', 'memory_used_percent': '%', 'memory_free_percent': '%',
    'disk_free_percent': '%', 'process_count': 'procs', 'uptime': 'hours',
  };
  return new TelemetrySeries({
    profileId: profile.id, provider: profile.provider,
    label: profile.label || `System ${variable}`,
    sensorName: `Local system: ${variable}`,
    times: currentTimes(values.length), values,
    unit: unitMap[variable] || '', usedLiveData: true,
  });
}

export async function fetchPing(profile, maxPoints = 512, _timeout = 12) {
  const host = profile.host || profile.params?.host || '127.0.0.1';
  const count = profile.count || profile.params?.count || 6;
  const rawResults = getPingMs(host, count);
  const valid = rawResults.filter(r => r > 0 && Number.isFinite(r));

  // Normalize: 0ms → health 1.0, 200ms → health 0.0
  const values = valid.length > 0 ? valid.map(r => Math.min(300, r)) : [];

  return new TelemetrySeries({
    profileId: profile.id, provider: profile.provider,
    label: profile.label || `Ping ${host}`,
    sensorName: `TCP ping RTT to ${host}`,
    times: values.map((_, i) => `probe:${i + 1}`),
    values,
    unit: 'ms',
    usedLiveData: values.length > 0,
  });
}

export { getCpuLoad, getMemoryUsedPercent, getDiskFreePercent, getProcessCount, getUptime, getPingMs };
